dotnet-svcutil.xmlserializer Assembly Load Error

**Describe the bug**
The dotnet-svcutil.xmlserializer is encountering an error because it is trying to load an assembly that does not exist. C:\Users\Admin\.nuget\packages\system.private.servicemodel\4.7.0\ does not contain a runtimes folder.

**To Reproduce**
Steps to reproduce the behavior:
1. Follow the steps in this article https://docs.microsoft.com/en-us/dotnet/core/additional-tools/dotnet-svcutil.xmlserializer-guide

**Expected behavior**
Generation of a serialization assembly.

**Screenshots**
`Microsoft (R) Build Engine version 16.4.0+e901037fe for .NET Core
Copyright (C) Microsoft Corporation. All rights reserved.

  Restore completed in 23.16 ms for C:\Users\Admin\source\repos\SvcUtilXmlSerializePreGen\SvcUtilXmlSerializePreGen.csproj.
  Restore completed in 59 ms for C:\Users\Admin\source\repos\SvcUtilXmlSerializePreGen\SvcUtilXmlSerializePreGen.csproj.
  SvcUtilXmlSerializePreGen -> C:\Users\Admin\source\repos\SvcUtilXmlSerializePreGen\bin\Debug\netstandard2.0\SvcUtilXmlSerializePreGen.dll
  Load Assembly From C:\Users\Admin\.nuget\packages\system.private.servicemodel\4.7.0\runtimes\win\lib\netstandard2.0\System.Private.ServiceModel.dll
EXEC : warning : Fail to load the assembly C:\Users\Admin\.nuget\packages\system.private.servicemodel\4.7.0\runtimes\win\lib\netstandard2.0\System.Private.ServiceModel.dll with the error Cannot load file C:\Users\Admin\.nuget\packages\system.private.servicemodel\4.7.0\runtimes\win\lib\netstandard2.0\System.Private.ServiceModel.dll as an Assembly. Check the FusionLogs for more Information. [C:\Users\Admin\source\repos\SvcUtilXmlSerializePreGen\SvcUtilXmlSerializePreGen.csproj]

  Microsoft (R) dotnet-svcutil.xmlserializer tool, Version 1.1.0.
  [Microsoft (R) Windows (R) Communication Foundation]
  Copyright (c) Microsoft Corporation.  All rights reserved.

  If you would like more help, type "svcutil -?"
EXEC : warning : Cannot load file C:\Users\Admin\.nuget\packages\system.private.servicemodel\4.7.0\runtimes\win\lib\netstandard2.0\System.Private.ServiceModel.dll as an Assembly. Check the FusionLogs for more Information. [C:\Users\Admin\source\repos\SvcUtilXmlSerializePreGen\SvcUtilXmlSerializePreGen.csproj]

      Could not load file or assembly 'C:\Users\Admin\.nuget\packages\system.private.servicemodel\4.7.0\runtimes\win\lib\netstandard2.0\System.Private.ServiceModel.dll'. The system cannot find the path specified.

C:\Users\Admin\.nuget\packages\dotnet-svcutil.xmlserializer\1.1.0\build\dotnet-svcutil.xmlserializer.targets(22,5): warning MSB3073: The command "dotnet svcutil.xmlserializer obj\Debug\netstandard2.0\SvcUtilXmlSerializePreGen.dll --quiet --out:obj\Debug\netstandard2.0\SvcUtilXmlSerializePreGen.XmlSerializers --smreference:"C:\Users\Admin\.nuget\packages\system.servicemodel.primitives\4.7.0\ref\netstandard2.0\System.ServiceModel.Primitives.dll"" exited with code 3. [C:\Users\Admin\source\repos\SvcUtilXmlSerializePreGen\SvcUtilXmlSerializePreGen.csproj]
C:\Users\Admin\.nuget\packages\dotnet-svcutil.xmlserializer\1.1.0\build\dotnet-svcutil.xmlserializer.targets(23,5): warning : Warning : Fail to generate the serializer for SvcUtilXmlSerializePreGen.dll. [C:\Users\Admin\source\repos\SvcUtilXmlSerializePreGen\SvcUtilXmlSerializePreGen.csproj]

Build succeeded.

EXEC : warning : Fail to load the assembly C:\Users\Admin\.nuget\packages\system.private.servicemodel\4.7.0\runtimes\win\lib\netstandard2.0\System.Private.ServiceModel.dll with the error Cannot load file C:\Users\Admin\.nuget\packages\system.private.servicemodel\4.7.0\runtimes\win\lib\netstandard2.0\System.Private.ServiceModel.dll as an Assembly. Check the FusionLogs for more Information. [C:\Users\Admin\source\repos\SvcUtilXmlSerializePreGen\SvcUtilXmlSerializePreGen.csproj]
EXEC : warning : Cannot load file C:\Users\Admin\.nuget\packages\system.private.servicemodel\4.7.0\runtimes\win\lib\netstandard2.0\System.Private.ServiceModel.dll as an Assembly. Check the FusionLogs for more Information. [C:\Users\Admin\source\repos\SvcUtilXmlSerializePreGen\SvcUtilXmlSerializePreGen.csproj]
C:\Users\Admin\.nuget\packages\dotnet-svcutil.xmlserializer\1.1.0\build\dotnet-svcutil.xmlserializer.targets(22,5): warning MSB3073: The command "dotnet svcutil.xmlserializer obj\Debug\netstandard2.0\SvcUtilXmlSerializePreGen.dll --quiet --out:obj\Debug\netstandard2.0\SvcUtilXmlSerializePreGen.XmlSerializers --smreference:"C:\Users\Admin\.nuget\packages\system.servicemodel.primitives\4.7.0\ref\netstandard2.0\System.ServiceModel.Primitives.dll"" exited with code 3. [C:\Users\Admin\source\repos\SvcUtilXmlSerializePreGen\SvcUtilXmlSerializePreGen.csproj]
C:\Users\Admin\.nuget\packages\dotnet-svcutil.xmlserializer\1.1.0\build\dotnet-svcutil.xmlserializer.targets(23,5): warning : Warning : Fail to generate the serializer for SvcUtilXmlSerializePreGen.dll. [C:\Users\Admin\source\repos\SvcUtilXmlSerializePreGen\SvcUtilXmlSerializePreGen.csproj]
    4 Warning(s)
    0 Error(s)

Time Elapsed 00:00:02.22`

**Additional context**
Repo: https://github.com/bobcat1506/SvcUtilXmlSerializePreGen

--- a/src/svcutilcore/files/dotnet-svcutil.xmlserializer.targets
+++ b/src/svcutilcore/files/dotnet-svcutil.xmlserializer.targets
@@ -9,10 +9,23 @@
     <_SerializationAssemblyDisabledWarnings>$(NoWarn);219;162;$(SerializationAssemblyDisabledWarnings)</_SerializationAssemblyDisabledWarnings>
   </PropertyGroup>
 
+  <UsingTask TaskName="Microsoft.NET.Build.Tasks.ResolvePackageAssets"
+            AssemblyFile="$(MicrosoftNETBuildTasksAssembly)" />
+
   <Target Name="SvcUtilGenerateSerializationAssembly" AfterTargets="Build">
+    <ResolvePackageAssets
+      ProjectAssetsFile="$(ProjectAssetsFile)"
+      ProjectAssetsCacheFile="$(ProjectAssetsCacheFile)"
+      ProjectPath="$(MSBuildProjectFullPath)"
+      TargetFrameworkMoniker="$(NuGetTargetMoniker)"
+      DotNetAppHostExecutableNameWithoutExtension="$(_DotNetAppHostExecutableNameWithoutExtension)">
+      <Output
+          TaskParameter="RuntimeAssemblies"
+          ItemName="WCFRuntimeAssembly"/>
+    </ResolvePackageAssets>
+
     <ItemGroup>
-      <_ReferenceAssembly Include="@(ReferencePath)"/>
-      <_ReferenceSMAssembly Include="@(_ReferenceAssembly)" Condition="$([System.String]::new('%(_ReferenceAssembly.Identity)').EndsWith('System.ServiceModel.Primitives.dll'))" />
+      <_ReferenceSMAssembly Include="@(WCFRuntimeAssembly)" />
     </ItemGroup>
     
     <Delete Condition="Exists('$(_SerializerDllIntermediateFolder)') == 'true'" Files="$(_SerializerDllIntermediateFolder)" ContinueOnError="true" />

--- a/src/svcutilcore/src/Microsoft/Tools/ServiceModel/SvcUtil/Options.cs
+++ b/src/svcutilcore/src/Microsoft/Tools/ServiceModel/SvcUtil/Options.cs
@@ -8,6 +8,7 @@ namespace Microsoft.Tools.ServiceModel.SvcUtil.XmlSerializer
     using System.Configuration;
     using System.Collections.Generic;
     using System.IO;
+    using System.Linq;
     using System.Reflection;
     using System.Runtime.InteropServices;
 
@@ -413,36 +414,29 @@ namespace Microsoft.Tools.ServiceModel.SvcUtil.XmlSerializer
 
             private void LoadSMReferenceAssembly()
             {
-                IList<string> referencedAssembliesArgs = _arguments.GetArguments(Options.Cmd.SMReference);
+                string smReferenceArg = _arguments.GetArgument(Options.Cmd.SMReference);
+                IList<string> referencedAssembliesArgs = smReferenceArg.Split(':').ToList();
                 if (referencedAssembliesArgs != null && referencedAssembliesArgs.Count > 0)
                 {
-                    string smassembly = referencedAssembliesArgs[0];
-                    // Allow the reference to be for System.Private.ServiceModel as well.
-                    if((smassembly.LastIndexOf("System.ServiceModel.Primitives", StringComparison.OrdinalIgnoreCase) == -1) && (smassembly.LastIndexOf("System.Private.ServiceModel", StringComparison.OrdinalIgnoreCase) == -1))
+                    string smassembly = "";
+                    string smpassembly = "";
+                    foreach (string path in referencedAssembliesArgs)
                     {
-                        ToolConsole.WriteError("Need to pass the right path of System.ServiceModel.Primitives for smreference parameter");
-                        throw new ArgumentException("Invalid smreference value");
+                        var file = new FileInfo(path);
+
+                        if (file.Name.Equals("System.ServiceModel.Primitives.dll", StringComparison.OrdinalIgnoreCase))
+                        {
+                            smassembly = path;
+                        }
+                        if (file.Name.Equals("System.Private.ServiceModel.dll", StringComparison.OrdinalIgnoreCase))
+                        {
+                            smpassembly = path;
+                        }
                     }
-
-                    string smpassembly = smassembly.Replace("System.ServiceModel.Primitives", "System.Private.ServiceModel");
-                    //for some lowercase path
-                    smpassembly = smpassembly.Replace("system.servicemodel.primitives", "system.private.servicemodel");
-
-                    int refplace = smassembly.LastIndexOf("ref");
-                    if(refplace > 0 )
+                    if ((string.IsNullOrEmpty(smassembly)) || (string.IsNullOrEmpty(smpassembly)))
                     {
-                        smassembly = smassembly.Remove(refplace, 3).Insert(refplace, "lib");
-                        refplace = smpassembly.LastIndexOf("ref");
-                        smpassembly = smpassembly.Remove(refplace, 3).Insert(refplace, "lib");
-                        int libplace = smpassembly.LastIndexOf("lib");
-                        if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
-                        {
-                            smpassembly = smpassembly.Remove(libplace, 3).Insert(libplace, @"runtimes\win\lib");
-                        }
-                        else
-                        {
-                            smpassembly = smpassembly.Remove(libplace, 3).Insert(libplace, @"runtimes/unix/lib");
-                        }
+                        ToolConsole.WriteError("Missing one or both of the paths for System.ServiceModel.Primitives and System.Private.ServiceModel");
+                        throw new ArgumentException("Invalid smreference value");
                     }
 
                     try
@@ -471,8 +465,8 @@ namespace Microsoft.Tools.ServiceModel.SvcUtil.XmlSerializer
                 }
                 else
                 {
-                    ToolConsole.WriteError("Need pass the System.ServiceModel.Primitive.dll through SM parameter");
-                    throw new ArgumentException("Need pass the System.ServiceModel.Primitive.dll through SM parameter");
+                    ToolConsole.WriteError("Need to pass the System.ServiceModel.Primitives.dll and the System.Private.ServiceModel.dll paths through the 'smreference' parameter.");
+                    throw new ArgumentException("Need to pass the System.ServiceModel.Primitives.dll and the System.Private.ServiceModel.dll paths through the 'smreference' parameter.");
                 }
             }
 

--- a/src/svcutilcore/src/dotnet-svcutil.xmlserializer.csproj
+++ b/src/svcutilcore/src/dotnet-svcutil.xmlserializer.csproj
@@ -100,5 +100,4 @@
     <Message Importance="high" Text="The value of property 'PsCmd' is: $(PsCmd)" />
     <Exec Command="$(PsCmd) -ExecutionPolicy UnRestricted -File $(WcfRootFolder)/src/svcutilcore/tools/scripts/UpdateSvcutilDotXmlSerializerPackage.ps1 $(PackageOutputPath)"/>
   </Target>
-
 </Project>

--- a/src/svcutilcore/tests/dotnet-svcutil.xmlserializer.IntegrationTests.csproj
+++ b/src/svcutilcore/tests/dotnet-svcutil.xmlserializer.IntegrationTests.csproj
@@ -20,7 +20,7 @@
     <!-- We're building netcoreap, run on the test CLI
          Reuse the same runtimeconfig that the tests use. -->
     <GeneratorRuntimeConfig>$(ArtifactsBinDir)\dotnet-svcutil.xmlserializer\$(Configuration)\$(TargetFramework)\dotnet-svcutil.xmlserializer.runtimeconfig.json</GeneratorRuntimeConfig>
-    <GeneratorCliPath>$(DOTNET_INSTALL_DIR)</GeneratorCliPath>
+    <GeneratorCliPath>$(DOTNET_INSTALL_DIR)/</GeneratorCliPath>
   </PropertyGroup>
 
   <PropertyGroup Condition="'$(TargetFramework)' != 'netcoreapp2.1'">
@@ -65,20 +65,27 @@
     </FindInList>
     
     <PropertyGroup>
-      <WcfSMOutputPath>$(ArtifactsBinDir)System.Private.ServiceModel\$(Configuration)\netstandard2.0</WcfSMOutputPath>
-      <ServiceModelAssemblyName>System.Private.ServiceModel.dll</ServiceModelAssemblyName>
-      <DotNetExe>$(GeneratorCliPath)\dotnet</DotNetExe>
+      <ServiceModelPrivateOutputPath>$(ArtifactsBinDir)System.Private.ServiceModel/$(Configuration)/netstandard2.0/</ServiceModelPrivateOutputPath>
+      <ServiceModelPrivateAssemblyName>System.Private.ServiceModel.dll</ServiceModelPrivateAssemblyName>
+      <ServiceModelPrimitivesOutputPath>$(ArtifactsBinDir)System.ServiceModel.Primitives.Facade/$(Configuration)/netstandard2.0/</ServiceModelPrimitivesOutputPath>
+      <ServiceModelPrimitivesAssemblyName>System.ServiceModel.Primitives.dll</ServiceModelPrimitivesAssemblyName>
+      <DotNetExe>$(GeneratorCliPath)dotnet</DotNetExe>
       <DotnetSvcutilXmlSerializerAssembly>$(OutputPath)dotnet-svcutil.xmlserializer.dll</DotnetSvcutilXmlSerializerAssembly>
       <InputTestAssembly>$(OutputPath)ScenarioTests.Common.Tests.dll</InputTestAssembly>
       <OutputGeneratedSource>$(OutputPath)ScenarioTests.Common.Tests.XmlSerializers.cs</OutputGeneratedSource>
       <CscOutputAssembly>$(OutputPath)ScenarioTests.Common.Tests.XmlSerializers.dll</CscOutputAssembly>
+      <ServiceModelPrivateAssembly>$(ServiceModelPrivateOutputPath)$(ServiceModelPrivateAssemblyName)</ServiceModelPrivateAssembly>
+      <ServiceModelPrimitivesAssembly>$(ServiceModelPrimitivesOutputPath)$(ServiceModelPrimitivesAssemblyName)</ServiceModelPrimitivesAssembly>
+      
     </PropertyGroup>
     <Message Text="Running Serialization Tool" Importance="normal" />
     
-    <Message Text="Run XmlSerializer: $(DotNetExe) $(DotnetSvcutilXmlSerializerAssembly) $(InputTestAssembly) --quiet --out:$(OutputGeneratedSource) --smreference:$(WcfSMOutputPath)\$(ServiceModelAssemblyName)  --reference=$(depRef)" Importance="high" />
-    <Message Text="$(ServiceModelAssemblyName) exists in $(WcfSMOutputPath)" Condition="Exists('$(WcfSMOutputPath)\$(ServiceModelAssemblyName)') == 'true'" Importance="high" />
-    <Error Text="$(ServiceModelAssemblyName) does not exist in $(WcfSMOutputPath)" Condition="Exists('$(WcfSMOutputPath)\$(ServiceModelAssemblyName)') != 'true'" />
-    <Exec Command="$(DotNetExe) $(DotnetSvcutilXmlSerializerAssembly) $(InputTestAssembly) --quiet --out:$(OutputGeneratedSource) --smreference:$(WcfSMOutputPath)\$(ServiceModelAssemblyName) --reference=$(depRef)" />
+    <Message Text="Run XmlSerializer: $(DotNetExe) $(DotnetSvcutilXmlSerializerAssembly) $(InputTestAssembly) --quiet --out:$(OutputGeneratedSource) --smreference:$(ServiceModelPrimitivesAssembly);$(ServiceModelPrivateAssembly) --reference=$(depRef)" Importance="high" />
+    <Message Text="$(ServiceModelPrivateAssemblyName) exists in $(ServiceModelPrivateOutputPath)" Condition="Exists('$(ServiceModelPrivateOutputPath)$(ServiceModelPrivateAssemblyName)') == 'true'" Importance="high" />
+    <Message Text="$(ServiceModelPrimitivesAssemblyName) exists in $(ServiceModelPrimitivesOutputPath)" Condition="Exists('$(ServiceModelPrimitivesOutputPath)$(ServiceModelPrimitivesAssemblyName)') == 'true'" Importance="high" />
+    <Error Text="$(ServiceModelPrivateAssemblyName) does not exist in $(ServiceModelPrivateOutputPath)" Condition="Exists('$(ServiceModelPrivateOutputPath)$(ServiceModelPrivateAssemblyName)') != 'true'" />
+    <Error Text="$(ServiceModelPrimitivesAssemblyName) does not exist in $(ServiceModelPrimitivesOutputPath)" Condition="Exists('$(ServiceModelPrimitivesOutputPath)$(ServiceModelPrimitivesAssemblyName)') != 'true'" />
+    <Exec Command="$(DotNetExe) $(DotnetSvcutilXmlSerializerAssembly) $(InputTestAssembly) --quiet --out:$(OutputGeneratedSource) --smreference:$(ServiceModelPrimitivesAssembly):$(ServiceModelPrivateAssembly) --reference=$(depRef)" />
     <Warning Condition="Exists('$(OutputGeneratedSource)') != 'true'" Text="Failed to generate $(OutputGeneratedSource)"/>
     <Csc Condition="Exists('$(OutputGeneratedSource)') == 'true'"
          OutputAssembly="$(CscOutputAssembly)"

